src/utils/ref_counted.h

```diff
@@ -0,0 +1,65 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_UTILS_REF_COUNTED_H_
+#define SRC_UTILS_REF_COUNTED_H_
+
+#include <iostream>
+
+#include "src/utils/ref_counter.h"
+
+namespace raksha {
+
+// Specialize this class for your type if you want custom reference counting.
+template <typename T>
+class RefCountManager {
+  static void Retain(T* ptr) { ptr->Retain(); }
+  static void Release(T* ptr) { ptr->Release(); }
+};
+
+// A helper class that allows a type `T` to be reference counted for an
+// `intrusive_ptr`.
+template <typename T>
+class RefCounted {
+ protected:
+  RefCounted() : count_(0) {}
+  RefCounted(const RefCounted& other) : RefCounted() {}
+  RefCounted(const RefCounted&& other) : RefCounted() {}
+
+  RefCounted& operator=(const RefCounted&) = delete;
+
+ private:
+  void Retain() { count_.FetchAdd(1); }
+
+  void Release() {
+    CHECK(count_ != 0) << "Reference count is already zero.";
+    size_t old = count_.FetchSub(1);
+    if (old == 1) {
+      delete this;
+    }
```

Comment:
In this context, you check whether the returned value `old` from `FetchSub` was 1, as you would if we were doing an atomic subtract operation. However, the `CHECK` assertion looks directly at the `count_` value, which does not show the same level of respect to the possibly-atomic abstraction. This wouldn't cause any problems, because we aren't doing anything atomic, but we should probably be consistent as to whether we are upholding an abstraction as if the count could be atomic.

Reply:
Good point. Actually, this check turned out to be redundant as `FetchSub` would already throw an error. I removed this CHECK and updated the test to reflect the new error message.